### DISCUSSION TITLE
Feature Request: ValueRelation construction from ParsedExpression

### DISCUSSION THREAD

**Discussion by anshuldata** [duckdb/duckdb#14689] [Ideas]
This is required to enhance duckdb-substrait extension to support VirtualTable values as rows of expression ([link](https://github.com/substrait-io/substrait/blob/a0a10bb39dae32a5564f4f37a4673068a08c8e91/proto/substrait/algebra.proto#L129)).  For e.g.  to represent ValueRelation for values of sql `select * from (values (1+1,2+2),(3+3,4))`. 
Existing ValueRelation constructors don't allow construction of valueRelation from expression ([link](https://github.com/duckdb/duckdb/blob/19864453f7d0ed095256d848b46e7b8630989bac/src/include/duckdb/main/relation/value_relation.hpp#L18-L21)). Request is to add another constructor to construct value relation from `vector<vector< ParsedExpression>>`

**Comment by anshuldata** [comment]
If this feature request is acceptable, I would like to work on this patch
Thanks

**Reply by Tishj** [reply]
The ParsedExpressions have to be foldable anyways, so I suggest to evaluate with `ExpressionExecutor::TryEvaluateScalar` to a Value and then create a ValueRelation like normal

**Reply by jacques-n** [reply]
This is for enhancing substrait <=> duckdb conversion. Our goal is to have a direct path that maintains symmetry and doesn't collapse user intent. (Note, I believe the substrait output from duckdb can produce constant expression trees so ideally input would be allowed for roundtrip purposes. 

Rough diff to support [can be seen here](https://github.com/sundeck-io/duckdb-substrait/commit/81ac8b4933d5d05b94f0d5694ab6877cc5d1317a).

**Reply by Tishj** [reply]
I dont think accepting a ParsedExpression is a good idea.

That is too broad, only ConstantExpression or other foldable expressions would be supported.

Instead of letting the constructor fail when it's given those unsupported expression types, I think it would be easier to evaluate the expressions into values and use the existing constructor.

**Reply by Tishj** [reply]
Actually, I'm working on the Python Expression API and found the current ValueRelation to be too limiting, I'm adding support for DefaultExpression and that wouldn't work with the current constructors, so I'll extend the ValueRelation to take expressions

**Reply by anshuldata** [reply]
one question @Tishj will new constructor take input as ParsedExpression or Expression ? If it is Expression, can you point me an API to which converts to Expression from ParsedExpression

**Reply by Tishj** [reply]
ParsedExpression

**Reply by anshuldata** [reply]
Perfect, Thanks

**Reply by Tishj** [reply]
PR is up, it's not actually solving my problem yet, because this still won't accept DefaultExpressions, as the ValueRelation gets bound before we are using it in the context of an insert statement
But never the less, I guess it can be useful to support expressions, the binder *should* catch it when expressions are used that can't be used in a ValueRelation

**Reply by anshuldata** [reply]
Thanks a lot @Tishj

**Comment by anshuldata** [comment]
imo this feature request can be closed @Tishj, Thanks

**Comment by jacques-n** [comment]
Adding a note here that I think this was merged in #14757

**Comment by anshuldata** [comment]
@Tishj 
I realized ParsedExpression constructor should have two variants similar to Value, one with ClientContext ([here](https://github.com/duckdb/duckdb/blob/ca5af32c331f9d5ea49f7158d5c83a47f25b8b79/src/include/duckdb/main/relation/value_relation.hpp#L18)) and one with RelationContextWrapper ([here](https://github.com/duckdb/duckdb/blob/ca5af32c331f9d5ea49f7158d5c83a47f25b8b79/src/include/duckdb/main/relation/value_relation.hpp#L22)). The one with RelationContextWrapper is missing (here) for ParsedExpression.
This is required to avoid taking client context lock (usage [here](https://github.com/substrait-io/duckdb-substrait-extension/blob/82bad97010067ccfd48bb6d996bd5f5591fcb85a/src/from_substrait.cpp#L627-L632))

this seems a straightforward change of adding another constructor, should I raise a PR ?

**Reply by anshuldata** [reply]
@Tishj 
Did you get a chance to check [this](https://github.com/duckdb/duckdb/discussions/14689#discussioncomment-11239740) ? I have captured changes required [this](https://github.com/anshuldata/duckdb/pull/2) draft PR in my fork.
Let me know wdyt ?

**Reply by Tishj** [reply]
That sounds fine yea, you can make a PR for that

**Reply by anshuldata** [reply]
Thanks @Tishj . Raised #14892 for same.
Thanks
